# Retail Sales Data Analysis & Customer Segmentation

## Level 2: Data Cleaning & Preparation

### Objective

The objective of this phase is to improve data quality and prepare the dataset for analysis.

This includes identifying and resolving missing values, duplicate records, invalid transactions, and data consistency issues.

By the end of this phase, the dataset will be cleaned, validated, and ready for feature engineering, exploratory analysis, and customer segmentation.

---

### Environment Setup

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Environment Ready")

Environment Ready


All required libraries were successfully imported and the analysis environment is ready.

---

### Dataset Loading

In [3]:
df = pd.read_csv("../data/Online Retail.csv", encoding="latin1")

The dataset was loaded successfully and is ready for the data cleaning process.

---

### Date Format Conversion

In [4]:
df["InvoiceDate"] = pd.to_datetime(
    df["InvoiceDate"],
    format="mixed"
)

The InvoiceDate column was converted to datetime format to support time-based analysis and feature engineering.

---

### Duplicate Records Assessment

In [5]:
df.duplicated().sum()

np.int64(5268)

A total of 5,268 duplicate records were identified within the dataset. These records could lead to inflated sales figures and inaccurate analytical results if not addressed.

---

### Duplicate Records Removal


In [6]:
df = df.drop_duplicates()

Duplicate transaction records were successfully removed to ensure that each transaction is represented only once within the dataset.

---

### Dataset Dimensions After Duplicate Removal

In [7]:
df.shape

(536641, 8)

After removing duplicate records, the dataset contains 536,641 rows and 8 columns, indicating a reduction in redundant transaction entries.

---

### Missing Values Assessment

In [8]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135037
Country             0
dtype: int64

Missing values were identified in the Description and CustomerID columns. CustomerID contains the majority of missing records and requires further attention before customer-level analysis can be performed.

---

### Missing Value Percentage Analysis

In [9]:
(df.isnull().sum()/len(df))*100

InvoiceNo       0.000000
StockCode       0.000000
Description     0.270945
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     25.163377
Country         0.000000
dtype: float64

Approximately 25.16% of records contain missing CustomerID values, while only 0.27% of records have missing product descriptions. Customer identifiers represent the primary data quality concern in the dataset.

---

### Missing Product Description Assessment

In [10]:
df["Description"].isnull().sum()

np.int64(1454)

A total of 1,454 transactions contain missing product descriptions, which may affect product-level analysis and reporting.

---

### Missing Product Description Handling

In [11]:
df = df.dropna(subset=["Description"])

Records with missing product descriptions were removed to maintain the quality and reliability of product-based analysis.

---

### Missing Customer Identifier Assessment

In [12]:
df["CustomerID"].isnull().sum()

np.int64(133583)

A total of 133,583 transactions do not contain customer identifiers, making customer-level analysis and segmentation impossible for these records.

---

### Missing Customer Identifier Handling

In [13]:
df = df.dropna(subset=["CustomerID"])

Transactions without customer identifiers were removed to ensure that all remaining records can be associated with a valid customer for CLV analysis and customer segmentation.

---

### Invalid Quantity Assessment

In [14]:
(df["Quantity"] <= 0).sum()

np.int64(8872)

A total of 8,872 transactions contain non-positive quantities, which likely represent cancelled orders, product returns, or invalid transaction records.

---

### Invalid Quantity Handling

In [15]:
df = df[df["Quantity"] > 0]

Transactions with non-positive quantities were removed to ensure that only valid sales transactions remain in the dataset.

---

### Invalid Unit Price Assessment

In [16]:
(df["UnitPrice"] <= 0).sum()

np.int64(40)

A small number of transactions (40 records) contain non-positive unit prices, which could negatively impact revenue calculations and business analysis.

---

### Invalid Unit Price Handling

In [17]:
df = df[df["UnitPrice"] > 0]

Transactions with non-positive unit prices were removed to maintain pricing consistency and support accurate revenue calculations.

---

### Missing Values Validation

In [18]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

All missing values have been successfully addressed. The cleaned dataset no longer contains null values in any column.

---

### Duplicate Records Validation

In [19]:
df.duplicated().sum()

np.int64(0)

No duplicate records remain in the dataset, confirming that duplicate removal was successfully completed.

---

### Transaction Quality Validation

In [20]:
(df["Quantity"] <= 0).sum(),
(df["UnitPrice"] <= 0).sum()

np.int64(0)

The dataset no longer contains transactions with non-positive quantities or non-positive unit prices, indicating successful transaction quality validation.

---

### Final Dataset Dimensions

In [21]:
df.shape

(392692, 8)

The final cleaned dataset contains 392,692 valid transaction records across 8 columns. The dataset is now suitable for feature engineering, exploratory analysis, customer analytics, and segmentation tasks.

---

### Export Cleaned Dataset

In [22]:
df.to_csv(
    "../outputs/cleaned_online_retail.csv",
    index=False
)

print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


The cleaned dataset was successfully exported and saved for use in subsequent phases of the project, ensuring reproducibility and workflow continuity.

---

### Summary

#### Key Cleaning Activities

The following data cleaning operations were performed to improve dataset quality and reliability:

- Converted the InvoiceDate column into datetime format.
- Removed 5,268 duplicate transaction records.
- Removed 1,454 records with missing product descriptions.
- Removed 133,583 records with missing customer identifiers.
- Removed 8,872 transactions containing non-positive quantities.
- Removed 40 transactions containing non-positive unit prices.
- Validated the dataset to ensure no remaining duplicates, missing values, or invalid transaction records.

---

#### Data Quality Improvements

Several data quality issues identified during the Data Understanding phase were successfully resolved.

The cleaned dataset no longer contains:

- Missing values
- Duplicate records
- Non-positive quantity values
- Non-positive unit price values

These improvements significantly enhance the reliability and accuracy of subsequent analysis.

---

#### Dataset Impact

The cleaning process resulted in a substantial improvement in dataset quality.

| Metric | Value |
|----------|----------|
| Original Records | 541,909 |
| Final Records | 392,692 |
| Total Records Removed | 149,217 |
| Final Columns | 8 |

The removed records primarily consisted of duplicate entries, transactions with missing customer information, incomplete product information, and invalid sales transactions.

---

#### Business Implications

The cleaned dataset now contains only valid and reliable customer purchase transactions.

This ensures more accurate results for:

- Revenue Analysis
- Product Performance Analysis
- Customer Lifetime Value (CLV) Analysis
- Customer Purchase Behavior Analysis
- Customer Segmentation using K-Means Clustering

By removing invalid and incomplete transactions, business insights generated from the dataset will be more trustworthy and actionable.

---

#### Conclusion

The Data Cleaning & Preparation phase successfully transformed the raw retail transaction dataset into a structured, high-quality, and analysis-ready dataset.

All major data quality issues were resolved, providing a strong foundation for advanced analytics and machine learning tasks.

---

#### Next Step

**Level 3: Feature Engineering**

The next phase will focus on creating business-oriented features such as:

- Revenue
- Year
- Month
- Day of Week
- Customer Purchase Frequency
- Average Order Value (AOV) etc.

These features will support exploratory analysis, customer analytics, and customer segmentation.